jailer.go

```diff
@@ -0,0 +1,372 @@
+	b.stdin = stdin
+	return b
+}
+
+// Build will build a jailer command.
+func (b JailerCommandBuilder) Build(ctx context.Context) *exec.Cmd {
+	cmd := exec.CommandContext(
+		ctx,
+		"jailer",
+		b.Args()...,
+	)
+
+	if stdin := b.Stdin(); stdin != nil {
+		cmd.Stdin = stdin
+	}
+
+	if stdout := b.Stdout(); stdout != nil {
+		cmd.Stdout = stdout
+	}
+
+	if stderr := b.Stderr(); stderr != nil {
+		cmd.Stderr = stderr
+	}
+
+	return cmd
+}
+
+// Jail will set up proper handlers and remove configuuration validation due to
+// stating of files
+func jail(ctx context.Context, m *Machine, cfg *Config) {
+	rootfs := ""
+	if len(cfg.JailerCfg.ChrootBaseDir) > 0 {
+		rootfs = filepath.Join(cfg.JailerCfg.ChrootBaseDir, "firecracker", cfg.JailerCfg.ID)
+	} else {
+		rootfs = filepath.Join(defaultJailerPath, cfg.JailerCfg.ID)
+	}
+
+	cfg.SocketPath = filepath.Join(rootfs, "api.socket")
+	m.cmd = JailerCommandBuilder{}.
+		WithID(cfg.JailerCfg.ID).
+		WithUID(IntValue(cfg.JailerCfg.UID)).
+		WithGID(IntValue(cfg.JailerCfg.GID)).
+		WithNumaNode(IntValue(cfg.JailerCfg.NumaNode)).
+		WithExecFile(cfg.JailerCfg.ExecFile).
+		WithChrootBaseDir(cfg.JailerCfg.ChrootBaseDir).
+		WithDaemonize(cfg.JailerCfg.Daemonize).
+		WithSeccompLevel(cfg.JailerCfg.SeccompLevel).
+		WithStdin(os.Stdin).
+		WithStdout(os.Stdout).
+		WithStderr(os.Stderr).
+		Build(ctx)
+
+	// ValidateCfgHandlerName handler needs to be removed due to checking for
+	// non-existant drive paths and socket path. Since the jailer uses relative
+	// links according to its rootfs then the user would also need those files
+	// relative to running this.
+	m.Handlers.Validation = m.Handlers.Validation.Remove(ValidateCfgHandlerName)
+	m.Handlers.FcInit = m.Handlers.FcInit.AppendAfter(CreateMachineHandlerName, Handler{
+		Name: "fcinit.CopyFilesToRootFS",
+		Fn: func(ctx context.Context, m *Machine) error {
+
+			// copy kernel image to root fs
+			kernelImageFileName := filepath.Base(m.cfg.KernelImagePath)
+			if err := linkFileToRootFS(
+				m.cfg.JailerCfg,
+				filepath.Join(rootfs, "root", kernelImageFileName),
+				m.cfg.KernelImagePath,
+			); err != nil {
+				return err
+			}
+
+			var rootDrive models.Drive
+			rootDriveIndex := 0
+			for i, drive := range m.cfg.Drives {
+				if BoolValue(drive.IsRootDevice) {
+					rootDrive = drive
+					rootDriveIndex = i
+					break
+				}
+			}
+
+			rootHostPath := StringValue(rootDrive.PathOnHost)
+			// copy root drive to root fs
+			rootdriveFileName := filepath.Base(rootHostPath)
+			if err := linkFileToRootFS(
+				m.cfg.JailerCfg,
+				filepath.Join(rootfs, "root", rootdriveFileName),
+				rootHostPath,
+			); err != nil {
+				return err
+			}
+
+			m.cfg.Drives[rootDriveIndex].PathOnHost = &rootdriveFileName
+			m.cfg.KernelImagePath = kernelImageFileName
+			return nil
+		},
+	})
+}
+
+func linkFileToRootFS(cfg JailerConfig, dst, src string) error {
+	if err := os.Link(src, dst); err != nil {
```

Comment:
If it's a plain file but on a different filesystem we could use a bind-mount instead of copying.

Reply:
Actually, I think we always need to copy. The file contains a filesystem. Unless we're doing something with a squashfs+read/write upper layer (which we're not, currently), this file can't safely be shared between concurrent VMs.

Reply:
Do we want to make the copying part of the SDK or make that requirement of non-concurrent reuse a thing exposed to the client?  I'm asking because we want to avoid copying when we're doing this with firecracker-containerd.